src/main/java/com/ge/predix/uaa/token/lib/FastTokenServices.java

```diff
@@ -128,8 +144,8 @@ public OAuth2Authentication loadAuthentication(final String accessToken) throws
 
         verifyIssuer(iss);
 
+        SignatureVerifier verifier = this.tokenKeys.getIfPresent(iss);
```

Comment:
@ben-manes could you please elaborate on the issue with getIfPresent()...

Reply:
`getIfPresent()` will perform the lookup, which may be null. Then the thread computes the value, which is expensive (hence the cache) and inserts it. This allows for a race where two threads query for the key, compute, and insert. This is known as a [cache stampede](https://en.wikipedia.org/wiki/Cache_stampede).

`get(key, function)` will atomically compute the value when absent. This acquires a per-entry lock so that only one thread does the work, subsequent calls block, and all receive the result. This is known as [memoization](https://en.wikipedia.org/wiki/Memoization).

Caffeine's `get` will perform a lock-free `getIfPresent` followed by a blocking `computeIfAbsent`, which avoids lock contention and duplicate computation work. We try to promote this style (hence `get` is shorter) and offer `LoadingCache` with additional functionality if you provide the computation function upfront.

Reply:
Thank you Ben :)